collections/src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -0,0 +1,519 @@
+    * @throws NoSuchElementException when empty
+    * @return
+    */
+  def removeHead(resizeInternalRepr: Boolean = false): A =
+    if (isEmpty) throw new NoSuchElementException(s"empty collection") else removeHeadAssumingNonEmpty(resizeInternalRepr)
+
+  @inline private[this] def removeHeadAssumingNonEmpty(resizeInternalRepr: Boolean = false): A = {
+    val elem = array(start)
+    array(start) = null
+    start = start_+(1)
+    if (resizeInternalRepr) resize(length)
+    elem.asInstanceOf[A]
+  }
+
+  /**
+    *
+    * @param resizeInternalRepr If this is set, resize the internal representation to reclaim space once in a while
+    * @return
+    */
+  def removeLastOption(resizeInternalRepr: Boolean = false): Option[A] =
+    if (isEmpty) None else Some(removeLastAssumingNonEmpty(resizeInternalRepr))
+
+  /**
+    * Unsafely remove the last element (throws exception when empty)
+    * See also removeLastOption()
+    *
+    * @param resizeInternalRepr If this is set, resize the internal representation to reclaim space once in a while
+    * @throws NoSuchElementException when empty
+    * @return
+    */
+  def removeLast(resizeInternalRepr: Boolean = false): A =
+    if (isEmpty) throw new NoSuchElementException(s"empty collection") else removeLastAssumingNonEmpty(resizeInternalRepr)
+
+  @inline private[this] def removeLastAssumingNonEmpty(resizeInternalRepr: Boolean = false): A = {
+    end = end_-(1)
+    val elem = array(end)
+    array(end) = null
+    if (resizeInternalRepr) resize(length)
+    elem.asInstanceOf[A]
+  }
+
+  /**
+    * Remove all elements from this collection and return the elements while emptying this data structure
+    * @return
+    */
+  def removeAll(): strawman.collection.Seq[A] = {
+    val elems = strawman.collection.Seq.newBuilder[A]
+    elems.sizeHint(length)
+    while(nonEmpty) {
+      elems += removeHeadAssumingNonEmpty()
+    }
+    elems.result()
+  }
+
+  /**
+    * Returns and removes all elements from the left of this queue which satisfy the given predicate
+    *
+    *  @param f   the predicate used for choosing elements
+    *  @return
+    */
+  def removeHeadWhile(f: A => Boolean): strawman.collection.Seq[A] = {
+    val elems = strawman.collection.Seq.newBuilder[A]
+    while(headOption.exists(f)) {
+      elems += removeHeadAssumingNonEmpty()
+    }
+    elems.result()
+  }
+
+  /**
+    * Returns and removes all elements from the right of this queue which satisfy the given predicate
+    *
+    *  @param f   the predicate used for choosing elements
+    *  @return
+    */
+  def removeLastWhile(f: A => Boolean): strawman.collection.Seq[A] = {
+    val elems = strawman.collection.Seq.newBuilder[A]
+    while(lastOption.exists(f)) {
+      elems += removeLastAssumingNonEmpty()
+    }
+    elems.result()
+  }
+
+  /**
+    * Returns the top (front) element of this queue (without removing it) and returns it (None if empty)
+    *
+    * @return
+    */
+  def peek: Option[A] = headOption
```

Comment:
You should make it `final` and `@inline` and indicate in the documentation that it is equivalent to `headOption`. BTW, is it really important to be able to write `peek` instead of `headOption`? I would stay with just `headOption` since this is what we already have.